### PR TITLE
Use string processing instead of regex

### LINKED ISSUE #13: Parse the string instead of the current regex mess
Regexes doesn't scale for this. Would be much easier to just iterate through the string and keep some state.

Something like: https://github.com/sindresorhus/strip-css-comments/blob/master/index.js

Should be something that can be shared with https://github.com/sindresorhus/to-double-quotes

### PR DESCRIPTION
**Pull request by ruipneves**
I saw this issue and decided to fork and give it a try, mainly because regular expressions can become really messy and confusing especially if things scale. The string processing suggested by you, was definitely a good idea to straighten things up.

The solution that I found for this problem was to iterate through the string maintaining a state, which is defined by the betweenQuotes variable. It is used to know when the processing is happening between two matching quotes and when the closing quote is found. While processing between matching quotes, escape and backslashes characters are processed in order to know when to escape a single quote or simply to escape the backslash.

What do you guys think?
Any suggestions?! Feel free to give me your advice. 

TODO:

- [ ] It would be interesting to add options in order to make the other way possible (single quotes -> double quotes).

---

Fixes #13

### CODE REVIEW

**Comment by satazor** [comment]
@ruipneves Could do update this PR to obey the code style? See: https://travis-ci.org/sindresorhus/to-single-quotes/jobs/193572769

If you are using atom or sublime search for `linter-xo` plugins, otherwise you may manually fix the issues by running `xo` manually inside `node_modules`.

**Comment by ruipneves** [comment]
I pushed the fix for the linter-xo but it has a problem for node version 0.10 and 0.12 as you can see in Travis log. This is due to the xo library being installed with the latest version which causes an error by using const which is not supported in versions 0.10 and 0.12.

**Comment by satazor** [comment]
@ruipneves You are right, we probably need to remove support for `0.10` and `0.12`. We should do this by removing them from `.travis.yml` and also forcing the [`engines`](https://docs.npmjs.com/files/package.json#engines) in package.json. @sindresorhus could you please give you feedback here?

**Comment by satazor** [comment]
@ruipneves I've added a few comments, sorry for the delay.

**Comment by ruipneves** [comment]
@satazor Thank you for your feedback. I just pushed the fix for the code styling errors. I still don't know how the strange identations got there. The only thing I did in the previous submission was applying the linter-xo fix to the code and manually fixing the errors that linter-xo couldn't fix automatically.

File: index.js [hunk 1]
```diff
@@ -1,53 +1,47 @@
...
-	for(x = 0; x < len; x++) {
-		//FOUND DOUBLE-QUOTE OR SINGLE-QUOTE
-		if(str[x] === '"' || str[x] === '\'') {
-			//IF NOT PROCESSING IN BETWEEN QUOTES
-			if(!betweenQuotes) {
+	for (x = 0; x < len; x++) {
+		// FOUND DOUBLE-QUOTE OR SINGLE-QUOTE
+		if (str[x] === '"' || str[x] === '\'') {
+			// IF NOT PROCESSING IN BETWEEN QUOTES
+			if (betweenQuotes === false) {
```
Comment (ruipneves) [inline index.js]: I had !betweenQuotes on the initial version but linter-xo pointed an error there when I ran the fix. Maybe the chain of issues below caused linter-xo to report that as an error as well?

**Comment by satazor** [comment]
@sindresorhus tests are failing on node < 4 because of xo, is it ok to remove support for older nodes right?

**Comment by SamVerschueren** [comment]
That or set XO to a fixed version. If I'm not mistaken, it should be `0.16.1`.

**Comment by sindresorhus** [comment]
> That or set XO to a fixed version. If I'm not mistaken, it should be 0.16.1.

👍 I will remove support for Node.js 0.10/0.12 after this PR is merged.

File: index.js [hunk 2]
```diff
@@ -1,16 +1,47 @@
...
-			.replace(/(^|[^\\])(\\+)'/g, '$1$2\\\'')
-			// escape single-quotes - round 1
-			.replace(/([^\\])'/g, '$1\\\'')
-			// escape single-quotes - round 2 (for consecutive single-quotes)
-			.replace(/([^\\])'/g, '$1\\\'')
-			// convert
-			.replace(/^"|"$/g, '\'');
-	});
+	var result = '';
+	var x;
```
Comment (sindresorhus) [inline index.js]: Use `i`, not `x`.

File: index.js [hunk 3]
```diff
@@ -1,16 +1,47 @@
...
+	var len = str.length;
```
Comment (sindresorhus) [inline index.js]: There's no point in caching the length. V8 is smart enough to do it for us.

File: index.js [hunk 4]
```diff
@@ -1,16 +1,47 @@
...
+				x++;
+			} else if (str[x + 1] === '"') {
+				// escape + double-quote
+				result += '"';
+				x++;
+			} else {
+				result += str[x];
+			}
+		} else if (str[x] === '\\' && str[x + 1] === '\\') {
+			// backslashes --> don't touch them
```
Comment (sindresorhus) [inline index.js]: Make it:

> don't touch backslashes

**Comment by sindresorhus** [comment]
Looks pretty good :)

---

> It would be interesting to add options in order to make the other way possible (single quotes -> double quotes).

Agreed. Not as a documented option, but to be able to reuse the logic here in [`to-double-quotes`](https://github.com/sindresorhus/to-double-quotes) without having to duplicate the code.

File: index.js (hunk 3 above)
Comment (ruipneves) [inline index.js]: @sindresorhus Thank you for the feedback! Just pushed the new version with the requested changes.

**Comment by ruipneves** [comment]
@sindresorhus Thank you for the feedback! Just pushed the new version with the requested changes.

**Comment by sindresorhus** [comment]
👍 Looks really good. Thank you @ruipneves :)

**Pull request merged**